ghcide/src/Development/IDE/Plugin/HLS.hs

```diff
@@ -436,35 +437,55 @@ makeCompletions :: [(PluginId, CompletionProvider IdeState)]
 makeCompletions sps lf ideState params@(CompletionParams (TextDocumentIdentifier doc) pos _context _mt)
   = do
       mprefix <- getPrefixAtPos lf doc pos
-      _snippets <- WithSnippets . completionSnippetsOn <$> getClientConfig lf
+      config <- getClientConfig lf
 
       let
           combine :: [CompletionResponseResult] -> CompletionResponseResult
-          combine cs = go (Completions $ List []) cs
-              where
-                  go acc [] = acc
-                  go (Completions (List ls)) (Completions (List ls2):rest)
-                      = go (Completions (List (ls <> ls2))) rest
-                  go (Completions (List ls)) (CompletionList (CompletionListType complete (List ls2)):rest)
-                      = go (CompletionList $ CompletionListType complete (List (ls <> ls2))) rest
-                  go (CompletionList (CompletionListType complete (List ls))) (CompletionList (CompletionListType complete2 (List ls2)):rest)
-                      = go (CompletionList $ CompletionListType (complete || complete2) (List (ls <> ls2))) rest
-                  go (CompletionList (CompletionListType complete (List ls))) (Completions (List ls2):rest)
-                      = go (CompletionList $ CompletionListType complete (List (ls <> ls2))) rest
-          makeAction (pid,p) = do
+          combine cs = go True mempty cs
+
+          go !comp acc [] =
+            CompletionList (CompletionListType comp (List $ DList.toList acc))
+          go comp acc (Completions (List ls) : rest) =
+            go comp (acc <> DList.fromList ls) rest
+          go comp acc (CompletionList (CompletionListType comp' (List ls)) : rest) =
+            go (comp && comp') (acc <> DList.fromList ls) rest
+
+          -- | Process a list of completion providers until we reach a max number of results
+          makeAction ::
+            Int ->
+            [(PluginId, CompletionProvider IdeState)] ->
+            IO [Either ResponseError CompletionResponseResult]
+          makeAction 0 _ = return []
+          makeAction _ [] = return []
+          makeAction n ((pid, p) : rest) = do
             pluginConfig <- getPluginConfig lf pid
-            if pluginEnabled pluginConfig plcCompletionOn
+            results <- if pluginEnabled pluginConfig plcCompletionOn
```

Comment:
It looks like we're generating the completions serially. That seems a bit sad. Why not do this step in parallel? I appreciate it makes it harder to put them back together, but I would have thought it was worth it, given how time-sensitive completions are. (Not necessarily for this diff)

Reply:
That's a good point, I was under the same impression. Enforcing max completions will make this harder, so maybe it's worth doing it in this PR